src/net_processing.cpp

```diff
@@ -4277,15 +4307,25 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
             nodestate->m_last_block_announcement = GetTime();
         }
 
-        std::map<uint256, std::pair<NodeId, std::list<QueuedBlock>::iterator> >::iterator blockInFlightIt = mapBlocksInFlight.find(pindex->GetBlockHash());
-        bool fAlreadyInFlight = blockInFlightIt != mapBlocksInFlight.end();
-
         if (pindex->nStatus & BLOCK_HAVE_DATA) // Nothing to do here
             return;
 
+        auto range_flight = mapBlocksInFlight.equal_range(pindex->GetBlockHash());
+        size_t already_in_flight = std::distance(range_flight.first, range_flight.second);
+        bool requested_block_from_this_peer{false};
+
+        while (range_flight.first != range_flight.second) {
+            if (range_flight.first->second.first == pfrom.GetId()) {
+                requested_block_from_this_peer = true;
+                break;
+            }
+            range_flight.first++;
+        }
+        bool first_in_flight = already_in_flight == 0 || (already_in_flight == 1 && requested_block_from_this_peer);
```

Comment:
This notion of `first_in_flight` is not correct, I think.  Imagine we get a headers announcement from an outbound (non-hb) peer, and so we send a getdata(cmpctblock).  In the meantime, we get a compact block announcement from an inbound peer which fails to reconstruct.  Then the compact block comes back from the outbound peer; `first_in_flight` will be false because we think we have two requests outstanding.

Further down, when we decide whether to send a GETBLOCKTXN message to this non-HB outbound peer, we will choose not to, because `first_in_flight` is false and the peer is not already HB.  This seems like it would be a regression, as inbound peers would gain the ability to interfere with compact block relay happening with other peers just by making a compactblock announcement.

Reply:
Similarly here, can we use the insertion order of multimap to determine which peer was actually first to make the announcement?  If in the situation where an outbound (non-hb) peer is the first to announce that we still always get the block, then this new logic would not be making anything worse, which seems sufficient to me.

Reply:
> If in the situation where an outbound (non-hb) peer is the first to announce that we still always send the getblocktxn if compact blocks fail, then this new logic would not be making anything worse, which seems sufficient to me.

Exactly. Fixed, I think!